app/components/autocomplete/at_mention/at_mention.js

```diff
@@ -50,74 +50,74 @@ export default class AtMention extends Component {
+        if (requestStatus === RequestStatus.NOT_STARTED || requestStatus === RequestStatus.SUCCESS) {
+            const sections = [];
             if (isSearch) {
-                data = {members: membersInChannel.concat(membersOutOfChannel).sort(sortByUsername)};
+                const {teamMembers} = autocompleteUsers;
+                sections.push({
+                    id: 'mobile.suggestion.members',
+                    defaultMessage: 'Members',
+                    data: teamMembers,
+                    key: 'teamMembers'
+                });
             } else {
-                if (membersInChannel.length > 0) {
-                    data = Object.assign({}, data, {inChannel: membersInChannel});
+                const {inChannel, outChannel} = autocompleteUsers;
+                if (inChannel.length > 0) {
+                    sections.push({
+                        id: 'suggestion.mention.members',
+                        defaultMessage: 'Channel Members',
+                        data: inChannel,
+                        key: 'inChannel'
+                    });
                 }
                 if (this.checkSpecialMentions(matchTerm) && !isSearch) {
-                    data = Object.assign({}, data, {specialMentions: this.getSpecialMentions()});
+                    sections.push({
+                        id: 'suggestion.mention.special',
+                        defaultMessage: 'Special Mentions',
+                        data: this.getSpecialMentions(),
+                        key: 'special',
+                        renderItem: this.renderSpecialMentions
+                    });
                 }
-                if (membersOutOfChannel.length > 0) {
-                    data = Object.assign({}, data, {notInChannel: membersOutOfChannel});
+                if (outChannel.length > 0) {
+                    sections.push({
+                        id: 'suggestion.mention.nonmembers',
+                        defaultMessage: 'Not in Channel',
+                        data: outChannel,
+                        key: 'outChannel'
+                    });
                 }
             }
 
             this.setState({
-                active: data.hasOwnProperty('inChannel') || data.hasOwnProperty('specialMentions') || data.hasOwnProperty('notInChannel') || data.hasOwnProperty('members'),
-                dataSource: this.state.dataSource.cloneWithRowsAndSections(data)
+                sections
             });
         }
     }
 
-    filter = (profiles, matchTerm) => {
-        const {isSearch} = this.props;
-        return profiles.filter((p) => {
-            return ((p.id !== this.props.currentUserId || isSearch) && (
-                p.username.toLowerCase().includes(matchTerm) || p.email.toLowerCase().includes(matchTerm) ||
-                p.first_name.toLowerCase().includes(matchTerm) || p.last_name.toLowerCase().includes(matchTerm)));
-        });
+    keyExtractor = (item) => {
+        return item.id || item;
```

Comment:
When is this not just `item.id`? The fallback seems misleading since item is always an object